### ISSUE TITLE
The result of __toString is mistakenly assumed to be always lower-cased

### ISSUE THREAD

**Issue by morozov** [vimeo/psalm#7536]
https://psalm.dev/r/dafc6637b5

> ERROR: RedundantFunctionCall - 11:6 - The call to strtolower is unnecessary

The above error is unexpected.

**Comment by weirdan** [comment]
https://psalm.dev/r/dafc6637b5

**Comment by orklah** [comment]
Indeed, there's a flaw in the code that checks that the return of __toString can be contained in lowercase-string:
https://github.com/vimeo/psalm/blob/a435bc57b986258cb2e3f52d6c76fa8b470997bf/src/Psalm/Internal/Type/Comparator/AtomicTypeComparator.php#L507

The condition right here behaves the same for every subtype of TString (so lowercase-string, but also non-empty-string etc...)

What we do is to check if the class is Stringable or has a __toString method, and if so, Psalm considers whatever the __toString returns match the container.

This is not always true however. __toString may return an empty string that would not be contained in `non-empty-string`.

What would need to be done here is retrieve the return type of __toString and compare it with the container. If they match, then it's time to flag the property

**Comment by weirdan** [comment]
Additional test cases: https://psalm.dev/r/2f6f853cfd
